File: 0133-clone-graph/0133-clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

from typing import Optional
class Solution:
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        if not node: return None
        h = {}
        
        def f(n):
            if n in h:
                return h[n]
            
            copy = Node(n.val)
            h[n] = copy
            for nei in n.neighbors:
                copy.neighbors.append(f(nei))
                
            return copy
        
        return f(node)
        

File: 0133-clone-graph/0133-clone-graph.java
/*
// Definition for a Node.
class Node {
    public int val;
    public List<Node> neighbors;
    public Node() {
        val = 0;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val) {
        val = _val;
        neighbors = new ArrayList<Node>();
    }
    public Node(int _val, ArrayList<Node> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
}
*/

class Solution {
    Map<Node, Node> cloneMap;

    Node cloneHelper(Node node, Map<Node, Node> cloneMap) {
        if (cloneMap.containsKey(node)){
            return cloneMap.get(node);
        }

        Node copy = new Node(node.val);
        cloneMap.put(node, copy);
        for (Node nei: node.neighbors) {
            copy.neighbors.add(cloneHelper(nei, cloneMap));
        }

        return copy;
    }

    public Node cloneGraph(Node node) {
        if (node == null) return node;
        cloneMap = new HashMap<>();
        return cloneHelper(node, cloneMap);

    }
}